### DISCUSSION TITLE
Consider making the Base Library 's method names to begin with lower casing

### DISCUSSION THREAD

**Discussion by hannah23280** [dotnet/csharplang#6580] [Language Ideas]
Currently, C# built-in methods all begin with upper casing.   I not sure for others, but imo, this make code readability really drops.  Imagine reading an english sentence where each word is captialized.

Hopefully in future, C# engineer can really consider making all methods begin with lower casing instead, or perhaps use some sort of compiler tricks, where our source codes are lower casing, and the compiler will convert to upper casing during compilation. Just an idea.

**Comment by huoyaoyuan** [comment]
> C# built-in methods all begin with upper casing.

Strictly speaking, there's no "built-in" "methods" for C# at all. All of the methods are provided by the base library.

C# only provides built-in alias for certain types, which are in lower case (`int`, `string`) etc.

> Imagine reading an english sentence where each word is captialized.

Imagine a method name composed by many words. It's really awkward for every word except for the first one to be capitalized.

> or perhaps use some sort of compiler tricks, where our source codes are lower casing, and the compiler will convert to upper casing during compilation

Technically speaking, this is a feature request about *referencing members in different name than they were compiled into IL*. F# supports this, but C# tends not to use too much tricks to make it differ from IL.

If such feature were implemented, I'd like to reference BCL method names in non-English.

**Reply by hannah23280** [reply]
> > C# built-in methods all begin with upper casing.
> 
> Strictly speaking, there's no "built-in" "methods" for C# at all. All of the methods are provided by the base library.
>
You are right. Should say "methods provided by the base library"
> 
> C# only provides built-in alias for certain types, which are in lower case (`int`, `string`) etc.
> 
> > Imagine reading an english sentence where each word is captialized.
> 
> Imagine a method name composed by many words. It's really awkward for every word except for the first one to be capitalized.
> 
Maybe it's really subjective. To me, I don't feel it weird if method name composed by many words except first one captialized, as in "test".toUpperCase()

**Reply by kant2002** [reply]
> Technically speaking, this is a feature request about referencing members in different name than they were compiled into IL. F# supports this, but C# tends not to use too much tricks to make it differ from IL.

Can you elaborate more on this? What you have in mind? Can I provide aliases for class name and methods?

**Comment by DavidArno** [comment]
> Imagine reading an english sentence where each word is captialized.

I Can Easily Imagine It. It Looks Odd As Wrong Information Is Being Conveyed, But It's Still Readable.

In English, some words - such as English itself - are capitalized mid-sentence. And despite all of this, C# isn't English; it's a programming language. The rules of one do not really affect the readability of the other.

> If such feature were implemented, I'd like to reference BCL method names in non-English.

Being English myself, I'd love the idea of being able to reference BCL items in English, rather than American. "Color" and "Math" could be hidden behind "Colour" and "Maths", and all would be good. 😜

**Reply by hannah23280** [reply]
> > Imagine reading an english sentence where each word is captialized.
> 
> I Can Easily Imagine It. It Looks Odd As Wrong Information Is Being Conveyed, But It's Still Readable.
> 
> In English, some words - such as English itself - are capitalized mid-sentence. And despite all of this, C# isn't English; it's a programming language. The rules of one do not really affect the readability of the other.
>

You are not wrong to say the above. But for me, coming from Java background (before which i also learn C# before), when i compare both casing convention, I find Java somehow more easily readable when the method first letter is small casing.   But if one have not learn Java or some other language similar to Java in terms of the casing before, one might not be able to feel the difference.

**Reply by spydacarnage** [reply]
As somebody who looked at Java before coming to C#, I can say I'm the other way around - the camel case method names drove me crazy and C#'s Pascal casing feels more natural.

Which shows that this is entirely a case of personal preference, and thus not something which will strongly sway the library designers.

Also, adding in the fact that the C# naming conventions do use camel case for other things, it would get awkward very quickly.

**Reply by hannah23280** [reply]
> As somebody who looked at Java before coming to C#, I can say I'm the other way around - the camel case method names drove me crazy and C#'s Pascal casing feels more natural.
> 
> Which shows that this is entirely a case of personal preference, and thus not something which will strongly sway the library designers.
> 
> Also, adding in the fact that the C# naming conventions do use camel case for other things, it would get awkward very quickly.

Glad to find someone who done both languages, and able to hear the opposite opinion.

**Reply by HaloFour** [reply]
I, for one, much prefer K&R braces common in Java compared to Allman braces which have largely been adopted in C#.  It took me a little while to get used to Java's conventions also coming from a C# and COM background but you do eventually get used to them and will find nearly every language has its own different conventions.

**Comment by HaloFour** [comment]
C# doesn't care about the casing of method names.  You can follow any convention you wish.

However, the convention that the BCL uses (and has always used) is that public method names adhere to Pascal case.  C# calling conventions are case sensitive, so of course you must call those methods using Pascal case.  C# isn't going to change to be case insensitive nor is it going to reinterpret the names of the identifiers you wish to use.

**Reply by hannah23280** [reply]
> C# doesn't care about the casing of method names. You can follow any convention you wish.
> 
> However, the convention that the BCL uses (and has always used) is that public method names adhere to Pascal case. C# calling conventions are case sensitive, so of course you must call those methods using Pascal case. C# isn't going to change to be case insensitive nor is it going to reinterpret the names of the identifiers you wish to use.

Yup.  For custom method names that we create ourselves, it does not have to follow convention.  I more of focusing on the API method names in the BCL (will update my post title to reflect this more clearly).  I also aware of the case sensitivity, which i have no issue with.    I don't expect C# to change case sensitivity or whatever, just trying to wish that C# can take into consideration whether to change all BCL method names' first letter to be lowercase.   
But of course, that would mean it would affect hundreds of existing C# application out that.

**Reply by theunrepentantgeek** [reply]
> But of course, that would mean it would affect hundreds of existing C# application out that.

You're wrong by several orders of magnitude.

There are at least millions of C# applications out there, and tens to hundreds of _billions_ of lines of existing code, if not more.

Let's do a back of the envelope calculation ...

Assuming there are only 10 billion lines of existing C# code that would need to be updated
... that only 50% of those lines of code would need to be changed 
... that there's only one change required per line
... that there's an automated tool to make all the changes
... that a competent developer doing a code review to catch errors would only need 4 seconds per change
... and that the developer is paid $50K per year to work 40 hours per week, with two weeks of vacation time each year.

This works out as taking 2,777 years of labour, for a total cost of $139 million dollars.

And that's a conservative estimate. All of my assumptions above are on the low side, some dramatically so.

**Reply by CyrusNajmabadi** [reply]
I'm not sure i've seen evidence of hundreds of C# applications in the first place.

**Reply by hannah23280** [reply]
> > But of course, that would mean it would affect hundreds of existing C# application out that.
> 
> You're wrong by several orders of magnitude.
> 
> There are at least millions of C# applications out there, and tens to hundreds of _billions_ of lines of existing code, if not more.
> 
> Let's do a back of the envelope calculation ...
> 
> Assuming there are only 10 billion lines of existing C# code that would need to be updated ... that only 50% of those lines of code would need to be changed ... that there's only one change required per line ... that there's an automated tool to make all the changes ... that a competent developer doing a code review to catch errors would only need 4 seconds per change ... and that the developer is paid $50K per year to work 40 hours per week, with two weeks of vacation time each year.
> 
> This works out as taking 2,777 years of labour, for a total cost of $139 million dollars.
> 
> And that's a conservative estimate. All of my assumptions above are on the low side, some dramatically so.

Haha...I realize you are reading too much into my word.  When i use the words "hundreds", i don't really mean mathematically hundreds as in between 100 to 999.  It's just a metaphorical way of saying " A LOT".

**Reply by mattwar** [reply]
I know I've written a few, so there has to be at least 10's of applications.

**Reply by TahirAhmadov** [reply]
> I'm not sure i've seen evidence of hundreds of C# applications in the first place.

I'm not sure I've seen evidence of the earth being an oblate spheroid in the first place. 
(Sorry, I'll see myself out)

**Comment by TahirAhmadov** [comment]
I'm pretty sure this cannot be done:
- prioritization - there are much more important fish to fry
- legacy - all existing code uses the Pascal case guideline, and there is no changing it at this point
- preference - I, among many many others, strongly prefer both the Pascal case and the brace indentation rule of C#, so there is that :)
- "reinterpretation"/aliasing - while I can see how it may be desirable for non-English speakers to be able to write in their native language, it's definitely not worth it just to change the casing of the first letter

To summarize, this falls under the same umbrella as the requests which appear from time to time to get rid of semicolons - these are stylistic choices that were made once and for all. Really, it's pretty easy to get used to them and it's absolutely not worth it to change the language to adapt to everybody's preferences.

**Comment by mattwar** [comment]
We cannot change the runtime to make casing changes.  However, you could use a different font in your editor that makes it look like there are no capital letters.

**Reply by hannah23280** [reply]
> We cannot change the runtime to make casing changes. However, you could use a different font in your editor that makes it look like there are no capital letters.

Thanks for this interesting trick. Never thought of using font.  But that would mean that ToUpperCase would now look like touppercase, which is not my intended.   Each word inside a multiple-words variable should still begin with captial letter, as in toUpperCase.

My original thinking is not to change the runtime, but the compiler. Programmer writes toUpperCase, and compiler can simply take the first letter and change it to upper casing in the IL.   Existing codes should not be affected, since it is already written upper case for first letter.  but of course, there can be new setting to allow developer to choose whether they want the compiler to do that or not.  But i know this might not be easy to implement,  and might take care of any side effect that arise from such change.

**Reply by yaakov-h** [reply]
There are probably fancy ways to use glyphs so that "\<space\>A" appears as "\<space\>a" or ".A" appears as ".a" and so on and so forth.
